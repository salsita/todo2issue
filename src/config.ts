import { config as loadDotenv } from 'dotenv'
import { resolve } from 'path'
import { readFile } from 'fs/promises'
import { promisify } from 'util'
import readJson from 'read-package-json'
import gitUrlParse from 'git-url-parse'

const readJsonAsync = promisify(readJson)

loadDotenv({
  path: resolve(process.cwd(), '.env')
})

export const GITHUB_TOKEN = process.env.GITHUB_TOKEN

export interface GitRepository {
  name: string
  owner: string
}

export interface Config {
  filePatterns: string[]
  repo: GitRepository
}

const defaultConfig: Partial<Config> = {
  filePatterns: ['**/*.[tj]s?(x)']
}

export async function readConfig (root: string): Promise<Config> {
  const {
    repository,
    todo2issue: {
      filePatterns = defaultConfig.filePatterns
    } = defaultConfig
  } = await readJsonAsync(resolve(root, 'package.json'))

  if (!repository?.url) {
    throw new Error(`'repository.url' missing in 'package.json'`)
  }

  const { name, owner } = gitUrlParse(repository?.url)

  if (!Array.isArray(filePatterns) || filePatterns.some(pattern => typeof pattern !== 'string')) {
    throw new Error(`'todo2issue.filePatterns' property is expected to be an array of glob patterns`)
  }
  return {
    filePatterns,
    repo: { name, owner }
  }
}
