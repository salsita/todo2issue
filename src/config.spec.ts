import { readConfig } from './config'
import { resolve } from 'path'

describe('readRepositoryConfig', () => {
  it('reads', async () => {
    const { githubToken, ...rest } = await readConfig(resolve(__dirname, '../sample'))
    expect(githubToken?.length).toBeGreaterThan(5)
    expect(rest).toMatchSnapshot()
  })
})
