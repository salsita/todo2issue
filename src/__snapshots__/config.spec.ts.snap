// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readRepositoryConfig reads 1`] = `
Object {
  "authorsByEmail": Object {
    "jirist@salsitasoft.com": "goce-cz",
  },
  "branch": "master",
  "filePatterns": Array [
    "sample/**/*.ts?(x)",
    "sample/**/*.js?(x)",
  ],
  "issueLabel": "TODO",
  "repo": Object {
    "name": "todo2issue",
    "owner": "salsita",
  },
}
`;
